src/composer.ts

```diff
@@ -464,14 +554,87 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
      * @param middleware The middleware to register
      */
     callbackQuery(
-        trigger: MaybeArray<string | RegExp>,
-        ...middleware: Array<Middleware<Filter<C, "callback_query:data">>>
-    ): Composer<Filter<C, "callback_query:data">> {
-        const trg = triggerFn(trigger);
-        return this.on("callback_query:data").filter(
-            (ctx) => match(ctx, ctx.callbackQuery.data, trg),
-            ...middleware,
-        );
+        trigger: MaybeNonEmptyArray<string>,
+        ...middleware: Array<
+            Middleware<FilteredMatchContext<C, string, "callback_query:data">>
+        >
+    ): Composer<FilteredMatchContext<C, string, "callback_query:data">>;
+    callbackQuery(
+        trigger: MaybeNonEmptyArray<RegExp>,
+        ...middleware: Array<
+            Middleware<
+                FilteredMatchContext<C, RegExpMatchArray, "callback_query:data">
+            >
+        >
+    ): Composer<
+        FilteredMatchContext<C, RegExpMatchArray, "callback_query:data">
+    >;
+    callbackQuery(
+        trigger: MustMixedNonEmptyArrayType<string, RegExp>,
+        ...middleware: Array<
+            Middleware<
+                FilteredMatchContext<
+                    C,
+                    string | RegExpMatchArray,
+                    "callback_query:data"
+                >
+            >
+        >
+    ): Composer<
+        FilteredMatchContext<
+            C,
+            string | RegExpMatchArray,
+            "callback_query:data"
+        >
+    >;
+    callbackQuery(
+        trigger: MaybeNonEmptyArray<string | RegExp>,
+        ...middleware: Array<
+            Middleware<
+                FilteredMatchContext<
+                    C,
+                    string | RegExpMatchArray,
+                    "callback_query:data"
+                >
+            >
+        >
```

Comment:
Is there a way to make this less verbose? I'm as uncomfortable as you probably are with half of the code in this file being type annotations of function overloads.